File: SamSafeCSharp/wwwroot/original source sam-safe/state.js
////////////////////////////////////////////////////////////////////////////////
// State
///////////////////////////////////////////////////////////////////////////////
// This is free and unencumbered software released into the public domain.
//
// Anyone is free to copy, modify, publish, use, compile, sell, or
// distribute this software, either in source code form or as a compiled
// binary, for any purpose, commercial or non-commercial, and by any
// means.
//
// In jurisdictions that recognize copyright laws, the author or authors
// of this software dedicate any and all copyright interest in the
// software to the public domain. We make this dedication for the benefit
// of the public at large and to the detriment of our heirs and
// successors. We intend this dedication to be an overt act of
// relinquishment in perpetuity of all present and future rights to this
// software under copyright law.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
// OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
// OTHER DEALINGS IN THE SOFTWARE.
//
// For more information, please refer to <http://unlicense.org/>


var state =  { } ;

state.init = function(render, view, display, intents) {
    
    if (render !== undefined) { 
        state.view = view ; 
        state.display = display || view.display ;
    }
    
    if (intents !== undefined) { state.intents = intents ; }
    
    if (render !== null) { state.render = render ; }
    
} ;

// Derive the state representation as a function of the systen
// control state
state.representation = function(model,next) {
    var representation = 'oops... something went wrong, the system is in an invalid state' ;

    if (state.ready(model)) {
        representation = state.view.ready(model, state.intents) ;
    } 
    
    state.display(representation,next) ;
    
    // return allowed actions
    return ['edit', 'save', 'delete', 'cancel'] ;
} ;

// Derive the current state of the system
state.ready = function(model) {
   return true ;
} ;


// Next action predicate, derives whether
// the system is in a (control) state where
// a new (next) action needs to be invoked

state.nextAction = function(model) {} ;

state.render = function(model,next) {
    console.log('in render') ;
    state.representation(model,next) ;
    state.nextAction(model) ;
} ;

module.exports = state ;

File: SamSafeCSharp/Components/State.cs
﻿using System;
using System.Collections.Generic;

namespace SamSafeCSharp.Components
{
    public class State
    {
        internal string[] AllowedActions { get; set; }
        private View _view;
        private Action<string, Action<string>> _display;
        private Action<Model, Action<string>> _render; //todo _render never used??
        private Dictionary<string, string> _intents;


        public void Init(Action<Model, Action<string>> render, View view, Action<string, Action<string>> display = null, Dictionary<string, string> intents = null)
        {
            if (render != null)
            {
                this._view = view;
                this._display = display ?? view.Display;
            }

            if (intents != null)
            {
                this._intents = intents;
            }

            this._render = render ?? this.DefaultRender;
        }

        /// <summary>
        /// Derive the state representation as a function of the system control state
        /// </summary>
        public string[] Representation(Model model, Action<string> next)
        {
            var representation = "Oops... something went wrong, the system";
            if (this.Ready(model))
            {
                representation = this._view.Ready(model, this._intents);
            }
            this._display(representation, next);

            // return allowed actions
            return new string[] { "edit", "save", "delete", "cancel" };
        }

        /// <summary>
        /// Next action predicate, derives whether the system is in a (control) state 
        /// where a new (next) action needs to be invoked
        /// </summary>
        public void NextAction(Model model)
        {

        }

        /// <summary>
        /// Derive the current state of the system
        /// </summary>
        private bool Ready(Model model)
        {
            return true;
        }

        private void DefaultRender(Model model, Action<string> next)
        {
            this.Representation(model, next);
            this.NextAction(model);
        }

    }
}
